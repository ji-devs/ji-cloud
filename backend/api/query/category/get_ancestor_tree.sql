with recursive links as
                   (
                       select id,
                              parent_id
                       from category co
                       where id = any ($1::uuid[])
                       union all
                       select co.id,
                              co.parent_id
                       from category co
                                inner join links ct on (ct.parent_id = co.id)
                   )

select distinct id,
       category.parent_id,
       name,
       category.index,
       created_at,
       updated_at,
       (select count(*) from image_category where category_id = id)::int8 as "image_count!",
       (select count(*)::int8 from jig_category where category_id = id) as "jig_count!",
        user_scopes
from category
         inner join links using (id);
