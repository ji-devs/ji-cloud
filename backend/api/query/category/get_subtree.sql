with recursive path(id, index, parent_id) as (
    select id, ord, null::uuid
    from category
             inner join unnest(
            $1::uuid[]) with ordinality t(id, ord)
                        using (id)
    union all
    select c.id, c.index, p.id
    from path p
             inner join category c on (c.parent_id = p.id)
)
select distinct id as "id!",
       path.index::int2 as "index!",
       path.parent_id,
       name,
       created_at,
       updated_at,
       (select count(*) from image_category where category_id = id)::int8 as "image_count!",
       (select count(*)::int8 from jig_category where category_id = id) as "jig_count!",
       user_scopes

from path
         inner join category using (id);
