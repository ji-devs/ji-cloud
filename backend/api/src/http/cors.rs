use actix_web::http::{header, Method};
use core::config::CORS_ORIGINS;

pub fn get(local_insecure: bool) -> actix_cors::Cors {
    let mut cors = actix_cors::Cors::default()
        .supports_credentials()
        .allowed_methods(&[
            Method::GET,
            Method::POST,
            Method::PATCH,
            Method::DELETE,
            Method::PUT,
            Method::OPTIONS,
        ])
        .allowed_headers(&[
            header::AUTHORIZATION,
            header::CONTENT_TYPE,
            header::HeaderName::from_static("x-csrf"),
        ]);

    if local_insecure {
        cors = cors.allow_any_origin();
    } else {
        for origin in CORS_ORIGINS {
            cors = cors.allowed_origin(origin);
        }
    }

    cors
}
