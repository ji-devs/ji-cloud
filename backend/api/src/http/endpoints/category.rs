use actix_web::{
    web::{self, Data, Json, Query, ServiceConfig},
    HttpResponse,
};
use shared::{
    api::endpoints::{category, ApiEndpoint, PathParts},
    domain::category::{
        CategoryId, CategoryResponse, CategoryTreeScope, CreateCategoryRequest, GetCategoryRequest,
        NewCategoryResponse, UpdateCategoryRequest,
    },
};
use sqlx::PgPool;
use tracing::instrument;

use crate::{
    db, error,
    extractor::{ScopeManageCategory, TokenUserWithScope},
};

pub enum CreateError {
    ParentCategoryNotFound,
    InternalServerError(anyhow::Error),
}

impl<T: Into<anyhow::Error>> From<T> for CreateError {
    fn from(e: T) -> Self {
        Self::InternalServerError(e.into())
    }
}

impl Into<actix_web::Error> for CreateError {
    fn into(self) -> actix_web::Error {
        match self {
            Self::ParentCategoryNotFound => error::BasicError::with_message(
                http::StatusCode::NOT_FOUND,
                "Parent Category Not Found".to_owned(),
            )
            .into(),
            Self::InternalServerError(e) => crate::error::ise(e),
        }
    }
}

/// Get a tree of categories.
#[instrument(skip(db))]
async fn get_categories(
    db: Data<PgPool>,
    req: Option<Query<<category::Get as ApiEndpoint>::Req>>,
) -> actix_web::Result<Json<<category::Get as ApiEndpoint>::Res>, error::Server> {
    let req = req.map_or_else(GetCategoryRequest::default, Query::into_inner);

    let categories = match req.scope {
        Some(CategoryTreeScope::Descendants) if req.ids.is_empty() => {
            db::category::get_tree(&db).await?
        }
        Some(CategoryTreeScope::Ancestors) | None if req.ids.is_empty() => {
            db::category::get_top_level(&db).await?
        }
        Some(CategoryTreeScope::Descendants) => db::category::get_subtree(&db, &req.ids).await?,
        Some(CategoryTreeScope::Ancestors) => {
            db::category::get_ancestor_tree(&db, &req.ids).await?
        }
        None => db::category::get_exact(&db, &req.ids).await?,
    };

    Ok(Json(CategoryResponse { categories }))
}

/// Create a category.
async fn create_category(
    db: Data<PgPool>,
    _claims: TokenUserWithScope<ScopeManageCategory>,
    req: Json<<category::Create as ApiEndpoint>::Req>,
) -> actix_web::Result<HttpResponse, CreateError> {
    let CreateCategoryRequest { name, parent_id } = req.into_inner();

    let (id, index) = db::category::create(&db, &name, parent_id).await?;

    Ok(HttpResponse::Created().json(NewCategoryResponse { id, index }))
}

/// Update a category.
async fn update_category(
    db: Data<PgPool>,
    _claims: TokenUserWithScope<ScopeManageCategory>,
    req: Option<Json<<category::Update as ApiEndpoint>::Req>>,
    path: web::Path<CategoryId>,
) -> actix_web::Result<HttpResponse, error::CategoryUpdate> {
    let UpdateCategoryRequest {
        name,
        parent_id,
        index,
        user_scopes,
    } = req.map_or_else(Default::default, Json::into_inner);

    db::category::update(
        &db,
        path.into_inner(),
        parent_id,
        name.as_deref(),
        index.map(|it| it as i16),
        user_scopes,
    )
    .await?;

    Ok(HttpResponse::NoContent().finish())
}

/// Delete a category.
async fn delete_category(
    db: Data<PgPool>,
    _claims: TokenUserWithScope<ScopeManageCategory>,
    path: web::Path<CategoryId>,
) -> actix_web::Result<HttpResponse, error::Delete> {
    db::category::delete(&db, path.into_inner()).await?;

    Ok(HttpResponse::NoContent().finish())
}

pub fn configure(cfg: &mut ServiceConfig) {
    cfg.route(
        <category::Get as ApiEndpoint>::Path::PATH,
        category::Get::METHOD.route().to(get_categories),
    )
    .route(
        <category::Create as ApiEndpoint>::Path::PATH,
        category::Create::METHOD.route().to(create_category),
    )
    .route(
        <category::Update as ApiEndpoint>::Path::PATH,
        category::Update::METHOD.route().to(update_category),
    )
    .route(
        <category::Delete as ApiEndpoint>::Path::PATH,
        category::Delete::METHOD.route().to(delete_category),
    );
}
