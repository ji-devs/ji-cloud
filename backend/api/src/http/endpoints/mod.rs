pub mod additional_resource;
pub mod admin;
pub mod animation;
pub mod audio;
pub mod category;
pub mod image;
pub mod jig;
pub mod locale;
pub mod media;
pub mod meta;
pub mod module;
pub mod search;
pub mod session;
pub mod user;
