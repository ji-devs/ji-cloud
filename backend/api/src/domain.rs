use actix_http::body::BoxBody;
use actix_web::{
    cookie::{Cookie, SameSite},
    HttpResponse, Responder,
};

use crate::{
    db::{self, account::AccountMember},
    error,
};
use chrono::{DateTime, Utc};
use http::StatusCode;
use shared::domain::billing::AccountId;
use shared::domain::user::{UserId, UserScope};
use shared::domain::{
    category::{Category, CategoryId},
    session::AUTH_COOKIE_NAME,
};
use sqlx::PgPool;
use std::{cell::RefCell, convert::TryFrom, rc::Rc};
use std::{collections::HashMap, fmt};
use tracing::instrument;
use uuid::Uuid;

#[derive(Debug)]
pub struct RawCategory {
    pub id: Uuid,
    pub parent_id: Option<Uuid>,
    pub name: String,
    pub index: i16,
    pub created_at: DateTime<Utc>,
    pub updated_at: Option<DateTime<Utc>>,
    pub user_scopes: Vec<i16>,
}

#[derive(Debug, Clone)]
struct CategoryNode {
    pub id: Uuid,
    pub name: String,
    pub created_at: DateTime<Utc>,
    pub updated_at: Option<DateTime<Utc>>,
    pub user_scopes: Vec<i16>,
    pub children: Vec<Rc<RefCell<CategoryNode>>>,
}

impl From<CategoryNode> for Category {
    fn from(category_node: CategoryNode) -> Self {
        Self {
            id: CategoryId(category_node.id),
            name: category_node.name,
            created_at: category_node.created_at,
            updated_at: category_node.updated_at,
            user_scopes: {
                category_node
                    .user_scopes
                    .into_iter()
                    .map(|x| UserScope::try_from(x).expect("detected an invalid user scope"))
                    .collect()
            },
            children: {
                category_node
                    .children
                    .iter()
                    .map(|category| {
                        let category: CategoryNode = (**category).borrow().clone();
                        category.into()
                    })
                    .collect()
            },
        }
    }
}

#[instrument(skip_all)]
pub fn build_tree(categories: Vec<RawCategory>) -> Vec<Category> {
    let mut nodes: Vec<Rc<RefCell<CategoryNode>>> = Vec::new();
    let mut lookup: HashMap<Uuid, Rc<RefCell<CategoryNode>>> = HashMap::new();

    // Now we know this category exists in the lookup table
    for raw in categories.iter() {
        lookup.insert(
            raw.id.clone(),
            Rc::new(RefCell::new(CategoryNode {
                id: raw.id.clone(),
                name: raw.name.clone(),
                created_at: raw.created_at,
                updated_at: raw.updated_at,
                user_scopes: raw.user_scopes.clone(),
                children: Vec::new(),
            })),
        );
    }

    for raw in categories.iter() {
        let current_node = lookup
            .get(&raw.id)
            .expect("Category wasn't added to lookup map");
        match raw.parent_id {
            None => {
                // This is a root category, add it to the nodes list
                nodes.push(current_node.clone());
            }
            Some(parent_id) => {
                let parent_node = lookup
                    .get(&parent_id)
                    .expect("Parent caregory wasn't added to lookup map");
                let mut parent_node = (**parent_node).borrow_mut();
                parent_node.children.push(current_node.clone());
            }
        }
    }

    nodes
        .iter()
        .map(|node| {
            let category: CategoryNode = (**node).borrow().clone();
            let category: Category = category.into();
            category
        })
        .collect()
}

#[derive(Debug, Copy, Clone, sqlx::Type)]
#[repr(u16)]
pub enum RegistrationStatus {
    /// The user was just registered ([`POST /v1/user`](shared::api::endpoints::user::Register))
    ///
    /// This state is *skipped* via ouath (they're automatically [`Validated`](Self::Validated))
    /// In this state the user is allowed to:
    /// * Request a new verification email
    /// * Request a password reset(?)
    New = 0,

    /// The user has gone through OAuth and created a new user, or has finished email verification.
    ///
    /// In this state the user is allowed to:
    /// * Request a password reset
    /// * Complete registration via creating their profile
    /// * Delete their account
    Validated = 1,

    /// The user has *completely* finished signup, and their profile has been completed.
    ///
    /// In this state the user is allowed to do anything that their scopes allow.
    /// All users can:
    /// * Change their email (not implemented nor currently on the roadmap)
    /// * Change their password if logged in, reset their password if not.
    /// * Access the API
    /// * Change their profile details (not currently implemented)
    /// * Delete their account
    Complete = 2,
}

#[derive(Debug)]
pub struct NoContentClearAuth;

impl fmt::Display for NoContentClearAuth {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str("No Content")
    }
}

impl Responder for NoContentClearAuth {
    type Body = BoxBody;

    fn respond_to(self, _: &actix_web::HttpRequest) -> HttpResponse<BoxBody> {
        let mut cookie = Cookie::named(AUTH_COOKIE_NAME);
        cookie.set_max_age(time::Duration::seconds(0));
        cookie.set_http_only(true);
        cookie.set_same_site(SameSite::Strict);

        actix_web::HttpResponse::build(StatusCode::NO_CONTENT)
            .content_type("application/octet-stream")
            .cookie(cookie)
            .finish()
    }
}

#[derive(Debug)]
pub enum UserAuthorization {
    SystemAdministrator,
    AccountAdministrator,
    AccountMember,
}

impl UserAuthorization {
    pub fn is_system_administrator(&self) -> bool {
        matches!(self, UserAuthorization::SystemAdministrator)
    }

    pub fn test_authorized(&self, require_account_admin: bool) -> Result<(), error::Account> {
        if self.is_authorized(require_account_admin) {
            Ok(())
        } else {
            Err(error::Account::Forbidden)
        }
    }

    pub fn is_authorized(&self, require_account_admin: bool) -> bool {
        match self {
            UserAuthorization::AccountMember if require_account_admin => false,
            _ => true,
        }
    }
}

pub async fn user_authorization(
    db: &PgPool,
    user_id: &UserId,
    account_id: &AccountId,
) -> Result<UserAuthorization, error::Account> {
    Ok(
        if db::user::has_scopes(db, *user_id, &[UserScope::Admin]).await? {
            UserAuthorization::SystemAdministrator
        } else {
            match db::account::user_account_membership(db, user_id, account_id).await? {
                Some(AccountMember::Admin) => UserAuthorization::AccountAdministrator,
                Some(AccountMember::User) => UserAuthorization::AccountMember,
                None => return Err(error::Account::Forbidden),
            }
        },
    )
}
