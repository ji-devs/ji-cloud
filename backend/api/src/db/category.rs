use crate::{
    domain::{build_tree, RawCategory},
    error::{self, Delete},
};
use futures::TryStreamExt;
use shared::domain::category::{Category, CategoryId};
use shared::domain::user::UserScope;
use sqlx::{Executor, PgPool};
use std::convert::TryFrom;
use tracing::instrument;
use uuid::Uuid;

#[instrument(skip_all)]
pub async fn get_top_level(db: &sqlx::PgPool) -> anyhow::Result<Vec<Category>> {
    sqlx::query!(
        r#"
select id                                                                 as "id: CategoryId",
       name,
       created_at,
       updated_at,
       (select count(*)::int8 from image_category where category_id = id) as "image_count!",
       (select count(*)::int8 from jig_data_category where category_id = id) as "jig_count!",
       user_scopes
from category
where parent_id is null
order by index
 "#
    )
    .fetch(db)
    .map_ok(|it| Category {
        id: it.id,
        created_at: it.created_at,
        updated_at: it.updated_at,
        name: it.name,
        children: vec![],
        image_count: it.image_count as u64,
        jig_count: it.jig_count as u64,
        user_scopes: {
            let scopes = it
                .user_scopes
                .into_iter()
                .map(|x| UserScope::try_from(x).expect("failed to convert to UserScope"))
                .collect();

            scopes
        },
    })
    .try_collect()
    .await
    .map_err(Into::into)
}

#[instrument(skip(db))]
pub async fn get_exact(db: &sqlx::PgPool, ids: &[Uuid]) -> sqlx::Result<Vec<Category>> {
    sqlx::query!(
        //language=SQL
        r#"
select id                                                                 as "id!: CategoryId",
       name                                                               as "name!",
       created_at                                                         as "created_at!",
       updated_at,
       (select count(*)::int8 from image_category where category_id = id) as "image_count!",
       (select count(*)::int8 from jig_data_category where category_id = id)   as "jig_count!",
       user_scopes                                                        as "user_scopes!"
from category
         inner join unnest($1::uuid[]) with ordinality t(id, ord) USING (id)
order by t.ord
"#,
        ids
    )
    .fetch(db)
    .map_ok(|it| Category {
        id: it.id,
        name: it.name,
        created_at: it.created_at,
        updated_at: it.updated_at,
        children: vec![],
        image_count: it.image_count as u64,
        jig_count: it.jig_count as u64,
        user_scopes: {
            it.user_scopes
                .into_iter()
                .map(|x| UserScope::try_from(x).expect("failed to convert to UserScope"))
                .collect::<Vec<_>>()
        },
    })
    .try_collect()
    .await
}

#[instrument(skip(db))]
pub async fn get_subtree(db: &sqlx::PgPool, ids: &[Uuid]) -> sqlx::Result<Vec<Category>> {
    sqlx::query_file_as!(RawCategory, "query/category/get_subtree.sql", ids)
        .fetch_all(db)
        .await
        .map(build_tree)
}

#[instrument(skip_all)]
pub async fn get_tree(db: &sqlx::PgPool) -> sqlx::Result<Vec<Category>> {
    sqlx::query_as!(
        RawCategory,
        r#"
select id,
       parent_id,
       name,
       index,
       created_at,
       updated_at,
       (select count(*) from image_category where category_id = id)::int8 as "image_count!",
       (select count(*) from jig_data_category where category_id = id)::int8 as "jig_count!",
       user_scopes
from category
"#
    )
    .fetch_all(db)
    .await
    .map(build_tree)
}

#[instrument(skip_all)]
pub async fn get_ancestor_tree(db: &sqlx::PgPool, ids: &[Uuid]) -> sqlx::Result<Vec<Category>> {
    sqlx::query_file_as!(RawCategory, "query/category/get_ancestor_tree.sql", ids)
        .fetch_all(db)
        .await
        .map(build_tree)
}

#[instrument(skip(db))]
pub async fn create(
    db: &sqlx::PgPool,
    name: &str,
    parent_id: Option<CategoryId>,
) -> sqlx::Result<(CategoryId, u16)> {
    let res = sqlx::query!(
        r#"
insert into category (index, parent_id, name, user_scopes)
VALUES((select count(*)::int2 from category where parent_id is not distinct from $1), $1, $2, array[]::smallint[])
returning index, id"#,
        parent_id.map(|it| it.0),
        name,
    )
    .fetch_one(db)
    .await?;

    Ok((CategoryId(res.id), res.index as u16))
}

/// checks if moving the category with `id` to have `new_parent` as it's parent would create a cycle
///
/// A cycle is where there's a category that is it's own descendant.
async fn would_cycle(
    txn: &mut sqlx::PgConnection,
    id: Uuid,
    new_parent: Uuid,
) -> sqlx::Result<bool> {
    // trivially having `self` as your parent would cycle, as you would have a child that's yourself.
    if new_parent == id {
        return Ok(true);
    }

    sqlx::query!(
        r#"
with recursive cte(parent_id) as (
select parent_id from category where id = $1
union all
select c.parent_id from category c inner join cte on cte.parent_id = c.id
) select exists(select 1 from cte where parent_id = $2) as "would_cycle!"
    "#,
        new_parent,
        id
    )
    .fetch_one(txn)
    .await
    .map(|res| res.would_cycle)
}

enum UpdateLoopError {
    UpdateError(error::CategoryUpdate),
    Sqlx(sqlx::Error),
}

impl From<sqlx::Error> for UpdateLoopError {
    fn from(err: sqlx::Error) -> Self {
        Self::Sqlx(err)
    }
}

impl From<error::CategoryUpdate> for UpdateLoopError {
    fn from(err: error::CategoryUpdate) -> Self {
        Self::UpdateError(err)
    }
}

async fn update_slow(
    db: &sqlx::PgPool,
    id: Uuid,
    new_parent: Option<Option<Uuid>>,
    name: Option<&str>,
    index: Option<i16>,
) -> Result<(), UpdateLoopError> {
    let mut txn = db.begin().await?;

    txn.execute("set transaction isolation level repeatable read")
        .await?;

    let category_info = sqlx::query!(
        r#"
select parent_id, index from category where id = $1 for update
    "#,
        id
    )
    .fetch_optional(&mut txn)
    .await?
    .ok_or(error::CategoryUpdate::CategoryNotFound)?;

    if let Some(name) = name {
        #[allow(clippy::suspicious_else_formatting)]
        sqlx::query!("update category set name = $1 where id = $2", name, id)
            .execute(&mut txn)
            .await?;
    }

    let mut current_index = category_info.index;
    if let Some(parent_id) = new_parent {
        if parent_id != category_info.parent_id {
            // check that the new parent isn't a descendant (to avoid cycles)
            if let Some(new_parent) = parent_id {
                let would_cycle = would_cycle(&mut txn, id, new_parent).await?;

                if would_cycle {
                    return Err(error::CategoryUpdate::Cycle.into());
                }
            }

            // handle the new parent not existing and return `ParentCategoryNotFound`
            let res = sqlx::query!(
                r#"
update category
set parent_id = $1,
    updated_at = now(),
    index = (select count(*)::int2 from category where parent_id is not distinct from $1)
where id = $2
returning index
"#,
                parent_id,
                id
            )
            .fetch_optional(&mut txn)
            .await?
            .ok_or(error::CategoryUpdate::ParentCategoryNotFound)?;

            current_index = res.index;

            backshift(&mut txn, category_info.parent_id, category_info.index, None).await?;
        }
    }

    if let Some(new_index) = index {
        let current_parent = new_parent.unwrap_or(category_info.parent_id);
        if new_index < current_index {
            sqlx::query!(
                r#"
update category
set updated_at = now(),
    index = index + 1
where index >= $1 and index < $2 and parent_id is not distinct from $3
                "#,
                new_index,
                current_index,
                current_parent
            )
            .execute(&mut txn)
            .await?;
        }

        if new_index > current_index {
            backshift(&mut txn, current_parent, current_index, Some(new_index)).await?;
        }

        if new_index != current_index {
            sqlx::query!(
                r#"
update category
set updated_at = now(),
    index = least((select count(*)::int2 from category c where c.parent_id is not distinct from parent_id), $1)
where id = $2
"#,
                index,
                id
            )
            .execute(&mut txn)
            .await?;
        }
    }

    txn.commit().await?;

    Ok(())
}

#[instrument(skip(db))]
pub async fn update(
    db: &sqlx::PgPool,
    CategoryId(id): CategoryId,
    parent_id: Option<Option<CategoryId>>,
    name: Option<&str>,
    index: Option<i16>,
    user_scopes: Option<Vec<UserScope>>,
) -> Result<(), error::CategoryUpdate> {
    // fast track for if we're only updating the `name`:
    // the reasoning is due to an observation:
    // * we have to have retry logic for anything that involves transactions here (which updating the parent id and index _requires_)
    // * This is a single query, which can execute faster than opening a transaction and doing all the other stuff does.
    if parent_id.is_none() && index.is_none() {
        if let Some(name) = name {
            let rows_updated = sqlx::query!(
                "update category set name = $1, updated_at = now() where id = $2",
                name,
                id
            )
            .execute(db)
            .await?
            .rows_affected();

            match rows_updated {
                0 => return Err(error::CategoryUpdate::CategoryNotFound),
                1 => {}
                _ => unreachable!(),
            }
        }

        if let Some(user_scopes) = user_scopes {
            let scopes: Vec<i16> = user_scopes.iter().map(|x| *x as i16).collect();

            let rows_updated = sqlx::query!(
                "update category set user_scopes= $1, updated_at = now() where id = $2",
                scopes.as_slice(),
                id
            )
            .execute(db)
            .await?
            .rows_affected();

            match rows_updated {
                0 => return Err(error::CategoryUpdate::CategoryNotFound),
                1 => {}
                _ => unreachable!(),
            }
        }

        // Regardless of if the name is updated,
        // we know that neither parent_id nor index are going to be updated,
        // so we can just return here.
        return Ok(());
    }

    // note: at this point, at least one of `parent_id` or `index` are not null,
    // however, we don't know if we will end up updating either (they could be the same)

    let new_parent = parent_id.map(|id| id.map(|it| it.0));
    loop {
        match update_slow(db, id, new_parent, name, index).await {
            Ok(()) => return Ok(()),
            Err(UpdateLoopError::UpdateError(e)) => return Err(e),
            Err(UpdateLoopError::Sqlx(sqlx::Error::Database(e)))
                if e.code().as_deref() == Some("40001") => {}
            Err(UpdateLoopError::Sqlx(e)) => {
                return Err(error::CategoryUpdate::InternalServerError(e.into()))
            }
        }
    }
}

async fn backshift(
    txn: &mut sqlx::PgConnection,
    parent_id: Option<Uuid>,
    old_index: i16,
    new_index: Option<i16>,
) -> sqlx::Result<()> {
    sqlx::query!(
        r#"
update category
set index = index - 1, updated_at = now()
where index > $1 and index <= $2 is not false and parent_id is not distinct from $3
"#,
        old_index,
        new_index,
        parent_id
    )
    .execute(txn)
    .await
    .map(drop)
}

pub async fn delete(db: &PgPool, id: CategoryId) -> Result<(), Delete> {
    async fn inner(db: &PgPool, id: CategoryId) -> sqlx::Result<()> {
        let mut txn = db.begin().await?;

        txn.execute("set transaction isolation level repeatable read")
            .await?;

        let res = sqlx::query!(
            "delete from category where id = $1 returning index, parent_id",
            id.0
        )
        .fetch_optional(&mut txn)
        .await?;

        if let Some(res) = res {
            backshift(&mut txn, res.parent_id, res.index, None).await?;
        }

        txn.commit().await?;

        Ok(())
    }

    loop {
        match inner(db, id).await {
            Ok(()) => return Ok(()),
            Err(sqlx::Error::Database(e)) if e.code().as_deref() == Some("40001") => {}
            Err(e) => return Err(e.into()),
        }
    }
}
