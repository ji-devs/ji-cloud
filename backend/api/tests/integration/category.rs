use http::StatusCode;
use serde_json::json;
use shared::domain::category::{
    CategoryTreeScope, CreateCategoryRequest, GetCategoryRequest, NewCategoryResponse,
};
use uuid::Uuid;

use crate::{
    fixture::Fixture,
    helpers::{initialize_server, LoginExt},
};

#[actix_rt::test]
async fn create() -> anyhow::Result<()> {
    let app = initialize_server(&[Fixture::User], &[]).await;

    let port = app.port();

    let client = reqwest::Client::new();

    let resp = client
        .post(&format!("http://0.0.0.0:{}/v1/category", port))
        .login()
        .json(&CreateCategoryRequest {
            name: "One".to_owned(),
            parent_id: None,
        })
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::CREATED);

    let _body: NewCategoryResponse = resp.json().await?;

    app.stop(false).await;

    Ok(())
}

#[actix_rt::test]
async fn get() -> anyhow::Result<()> {
    let app = initialize_server(&[Fixture::User, Fixture::CategoryOrdering], &[]).await;

    let port = app.port();

    let client = reqwest::Client::new();

    let resp = client
        .get(&format!("http://0.0.0.0:{}/v1/category", port))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::OK);

    let body: serde_json::Value = resp.json().await?;

    app.stop(false).await;

    insta::assert_json_snapshot!(body);

    Ok(())
}

async fn get_nested_categories(query: &GetCategoryRequest) -> anyhow::Result<()> {
    let app = initialize_server(&[Fixture::User, Fixture::CategoryNesting], &[]).await;

    let port = app.port();

    let client = reqwest::Client::new();

    let resp = client
        .get(&format!("http://0.0.0.0:{}/v1/category", port))
        .query(query)
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::OK);

    let body: serde_json::Value = resp.json().await?;

    app.stop(false).await;

    insta::assert_json_snapshot!(body);

    Ok(())
}

#[actix_rt::test]
async fn nested_top_level() -> anyhow::Result<()> {
    get_nested_categories(&GetCategoryRequest::default()).await
}

#[actix_rt::test]
async fn nested_whole_tree() -> anyhow::Result<()> {
    get_nested_categories(&GetCategoryRequest {
        scope: Some(CategoryTreeScope::Decendants),
        ids: vec![],
    })
    .await
}

#[actix_rt::test]
async fn nested_overlapping() -> anyhow::Result<()> {
    get_nested_categories(&GetCategoryRequest {
        scope: Some(CategoryTreeScope::Decendants),
        ids: vec![
            "afbce03c-e90f-11ea-8281-cfde02f6b582".parse()?,
            "e315d3b2-e90f-11ea-8281-73cd69c14821".parse()?,
        ],
    })
    .await
}

#[actix_rt::test]
async fn nested_ancestors() -> anyhow::Result<()> {
    get_nested_categories(&GetCategoryRequest {
        scope: Some(CategoryTreeScope::Ancestors),
        ids: vec!["e315d3b2-e90f-11ea-8281-73cd69c14821".parse()?],
    })
    .await
}

#[actix_rt::test]
async fn nested_exact() -> anyhow::Result<()> {
    get_nested_categories(&GetCategoryRequest {
        scope: None,
        ids: vec![
            "afbce03c-e90f-11ea-8281-cfde02f6b582".parse()?,
            "01cff7d8-e910-11ea-8281-7f86c625a156".parse()?,
        ],
    })
    .await
}

#[actix_rt::test]
async fn upgdate_ordering() -> anyhow::Result<()> {
    let category_three = "81c4796a-e883-11ea-93f0-df2484ab6b11";

    let app = initialize_server(&[Fixture::User, Fixture::CategoryOrdering], &[]).await;

    let port = app.port();

    let client = reqwest::Client::new();

    let resp = client
        .patch(&format!(
            "http://0.0.0.0:{}/v1/category/{}",
            port, category_three
        ))
        .json(&json!({"index": 0}))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::NO_CONTENT);

    let resp = client
        .get(&format!("http://0.0.0.0:{}/v1/category", port))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::OK);

    let body: serde_json::Value = resp.json().await?;

    insta::assert_json_snapshot!(body, {".**.updated_at" => "[timestamp]"});

    let resp = client
        .patch(&format!(
            "http://0.0.0.0:{}/v1/category/{}",
            port, category_three
        ))
        .json(&json!({"index": 2}))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::NO_CONTENT);

    let resp = client
        .get(&format!("http://0.0.0.0:{}/v1/category", port))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::OK);

    let body: serde_json::Value = resp.json().await?;

    app.stop(false).await;

    insta::assert_json_snapshot!(body, {".**.updated_at" => "[timestamp]"});

    Ok(())
}

#[actix_rt::test]
async fn delete() -> anyhow::Result<()> {
    let app = initialize_server(&[Fixture::User, Fixture::CategoryOrdering], &[]).await;

    let port = app.port();

    let client = reqwest::Client::new();

    let resp = client
        .delete(&format!(
            "http://0.0.0.0:{}/v1/category/7fe19326-e883-11ea-93f0-5343493c17c4",
            port,
        ))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::NO_CONTENT);

    let resp = client
        .get(&format!("http://0.0.0.0:{}/v1/category", port))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::OK);

    let body: serde_json::Value = resp.json().await?;

    app.stop(false).await;

    insta::assert_json_snapshot!(body, {".**.updated_at" => "[timestamp]"});

    Ok(())
}

async fn update(id: Uuid, body: &serde_json::Value) -> anyhow::Result<()> {
    let app = initialize_server(&[Fixture::User, Fixture::CategoryOrdering], &[]).await;

    let port = app.port();

    let client = reqwest::Client::new();

    let resp = client
        .patch(&format!("http://0.0.0.0:{}/v1/category/{}", port, id))
        .json(&body)
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::NO_CONTENT);

    let resp = client
        .get(&format!(
            "http://0.0.0.0:{}/v1/category?scope=Decendants",
            port
        ))
        .login()
        .send()
        .await?
        .error_for_status()?;

    assert_eq!(resp.status(), StatusCode::OK);

    let body: serde_json::Value = resp.json().await?;

    app.stop(false).await;

    insta::assert_json_snapshot!(body, {".**.updated_at" => "[timestamp]"});

    Ok(())
}

#[actix_rt::test]
async fn update_parent() -> anyhow::Result<()> {
    update(
        "7fe19326-e883-11ea-93f0-5343493c17c4".parse()?,
        &json!({"parent_id": "81c4796a-e883-11ea-93f0-df2484ab6b11"}),
    )
    .await
}

#[actix_rt::test]
async fn update_reparent_move() -> anyhow::Result<()> {
    update(
        "7fe19326-e883-11ea-93f0-5343493c17c4".parse()?,
        &json!({"parent_id": (), "index": 0}),
    )
    .await
}

#[actix_rt::test]
async fn update_move() -> anyhow::Result<()> {
    update(
        "81c4796a-e883-11ea-93f0-df2484ab6b11".parse()?,
        &json!({"index": 1}),
    )
    .await
}

#[actix_rt::test]
async fn update_rename() -> anyhow::Result<()> {
    update(
        "81c4796a-e883-11ea-93f0-df2484ab6b11".parse()?,
        &json!({"name": "abc123"}),
    )
    .await
}
