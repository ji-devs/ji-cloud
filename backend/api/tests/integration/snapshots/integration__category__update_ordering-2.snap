---
source: tests/integration/category.rs
expression: body
---
{
  "categories": [
    {
      "id": "7e0cc6e2-e883-11ea-93f0-cb82f9b3da95",
      "user_scopes": [
        "Admin",
        "ManageImage",
        "AdminAsset"
      ],
      "name": "One",
      "created_at": "2020-08-27T16:36:44.133612Z",
      "updated_at": "[timestamp]"
    },
    {
      "id": "7fe19326-e883-11ea-93f0-5343493c17c4",
      "user_scopes": [
        "ManageCategory"
      ],
      "name": "Two",
      "created_at": "2020-08-27T16:36:47.206959Z",
      "updated_at": "[timestamp]"
    },
    {
      "id": "81c4796a-e883-11ea-93f0-df2484ab6b11",
      "user_scopes": [
        "Admin",
        "ManageAnimation"
      ],
      "name": "Three",
      "created_at": "2020-08-27T16:36:50.371709Z",
      "updated_at": "[timestamp]"
    }
  ]
}
